File: testing/suites/features/fortran/datatypes/ckpt_primitives.c
/**
 * Copyright (c) 2017 Leonardo A. Bautista-Gomez
 * All rights reserved
 *
 * @file   ckpt_primitives.c
 * @author Alexandre de Limas Santana (alexandre.delimassantana@bsc.es)
 * @date   August, 2020
**/
#include <fti.h>
#include <stdio.h>
#include <stdlib.h>

#define SIZE 5

// Patterns to initialize primitive data for testing
// Any pattern will do, these were selected empirically
#define C1_INIT 'A'             // Stored as its ASCII value, 65
#define LX_INIT 1               // Fortran 'true' maps to 1 in C
#define I1_INIT 127 - i         // Greater 8-bit number minus i
#define I2_INIT 32767 - i       // Greater 16-bit number minus i
#define I4_INIT 2147483647 - i  // Greater 32-bit number minus i
#define I8_INIT 2147483647 + i  // Greater 32-bit number plus i
#define R4_INIT 10 + i
#define R8_INIT 20 + i
#define R16_INIT 30 + i

typedef struct Complex4 {
    float r, i;
} Complex4;

typedef struct Complex8 {
    double r, i;
} Complex8;

typedef struct Complex16 {
    long double r, i;
} Complex16;

int main(int argc, char * argv[]) {
  char *configfile;
  int doInitData, rank, ret, i;

  /**
   * These are the variables for testing.
   * Their types should be verified in the meta-data files.
  **/
  char c1[SIZE];

  // Logicals are stored as integers in ANSI C, there are no booleans
  char i1[SIZE], l1[SIZE];
  short i2[SIZE], l2[SIZE];
  int i4[SIZE], l4[SIZE];
  long i8[SIZE], l8[SIZE];

  float r4[SIZE];
  double r8[SIZE];
  long double r16[SIZE];

  Complex4 cp4[SIZE];
  Complex8 cp8[SIZE];
  Complex16 cp16[SIZE];

  fti_id_t complex_4, complex_8, complex_16;

  if (argc < 2) {
    printf("Expected two arguments: configfile (str) and doInitData (bool)");
    return 1;
  }

  configfile = argv[1];
  doInitData = atoi(argv[2]);

  MPI_Init(&argc, &argv);
  FTI_Init(configfile, MPI_COMM_WORLD);
  MPI_Comm_rank(FTI_COMM_WORLD, &rank);

  /**
   * Conditionally initialize the data based on command-line arguments.
   * Any initialization will do as long as the verification repeats it.
   **/
  if (doInitData) {
    if (rank == 0) {
      printf("Application initializes its own data");
    }
    for (i = 0; i < SIZE; ++i) {
      c1[i] = C1_INIT;
      l1[i] = LX_INIT;
      l2[i] = LX_INIT;
      l4[i] = LX_INIT;
      l8[i] = LX_INIT;
      i1[i] = I1_INIT;
      i2[i] = I2_INIT;
      i4[i] = I4_INIT;
      i8[i] = I8_INIT;
      r4[i] = R4_INIT;
      r8[i] = R8_INIT;
      r16[i] = R16_INIT;
      cp4[i].i = i;
      cp4[i].r = i;
      cp8[i].i = i;
      cp8[i].r = i;
      cp16[i].i = i;
      cp16[i].r = i;
    }
  }
  // Initializing custom types
  complex_4 = FTI_InitComplexType("Complex4", sizeof(Complex4), NULL);
  FTI_AddSimpleField(complex_4, "r", FTI_SFLT, offsetof(Complex4, r));
  FTI_AddSimpleField(complex_4, "i", FTI_SFLT, offsetof(Complex4, i));
  complex_8 = FTI_InitComplexType("Complex8", sizeof(Complex8), NULL);
  FTI_AddSimpleField(complex_8, "r", FTI_SFLT, offsetof(Complex8, r));
  FTI_AddSimpleField(complex_8, "i", FTI_SFLT, offsetof(Complex8, i));
  complex_16 = FTI_InitComplexType("Complex16", sizeof(Complex16), NULL);
  FTI_AddSimpleField(complex_16, "r", FTI_SFLT, offsetof(Complex16, r));
  FTI_AddSimpleField(complex_16, "i", FTI_SFLT, offsetof(Complex16, i));

  FTI_Protect(0, c1, SIZE, FTI_CHAR);

  FTI_Protect(10, l1, SIZE, FTI_CHAR);
  FTI_Protect(11, l2, SIZE, FTI_SHRT);
  FTI_Protect(12, l4, SIZE, FTI_INTG);
  FTI_Protect(13, l8, SIZE, FTI_LONG);

  FTI_Protect(20, i1, SIZE, FTI_CHAR);
  FTI_Protect(21, i2, SIZE, FTI_SHRT);
  FTI_Protect(22, i4, SIZE, FTI_INTG);
  FTI_Protect(23, i8, SIZE, FTI_LONG);

  FTI_Protect(30, r4, SIZE, FTI_SFLT);
  FTI_Protect(31, r8, SIZE, FTI_DBLE);
  FTI_Protect(32, r16, SIZE, FTI_LDBE);

  FTI_Protect(40, cp4, SIZE, complex_4);
  FTI_Protect(41, cp8, SIZE, complex_8);
  FTI_Protect(42, cp16, SIZE, complex_16);

  /**
   * If the application initialized the data, do the following:
   * 1) Checkpoint the data using FTI_Checkpoint()
   * 2) Simulate a crash by not calling FTI_Finalize()
   * 3) Return 0 because everything went as planned
  **/
  if (doInitData != 0) {
    FTI_Checkpoint(1, 1);
    MPI_Finalize();
    return 0;
  } else {
    // If the application did not initialize the data, FTI does so.
    // In this case, we move on to check if the values are correct.
    FTI_Recover();
  }

  // Test if data was initialized following the same patterns
  // Checks if the IO library stored data in accordance to their binary format.
  ret = 1;
  for (i = 0; i < SIZE; ++i) {
    if (c1[i] != C1_INIT) {
      if (rank == 0)
        printf("character(1) was corrupted %c\n", c1[i]);
      goto end;
    }

    if (l1[i] != LX_INIT) {
      if (rank == 0)
        printf("logical(1) was corrupted %d\n", l1[i]);
      goto end;
    }
    if (l2[i] != LX_INIT) {
      if (rank == 0)
        printf("logical(2) was corrupted %d\n", l2[i]);
      goto end;
    }
    if (l4[i] != LX_INIT) {
      if (rank == 0)
        printf("logical(4) was corrupted %d\n", l4[i]);
      goto end;
    }
    if (l8[i] != LX_INIT) {
      if (rank == 0)
        printf("logical(8) was corrupted %ld\n", l8[i]);
      goto end;
    }

    if (i1[i] != I1_INIT) {
      if (rank == 0)
        printf("integer(1) was corrupted %d\n", i1[i]);
      goto end;
    }
    if (i2[i] != I2_INIT) {
      if (rank == 0)
        printf("integer(2) was corrupted %hd\n", i2[i]);
      goto end;
    }
    if (i4[i] != I4_INIT) {
      if (rank == 0)
        printf("integer(4) was corrupted %d\n", i4[i]);
      goto end;
    }
    if (i8[i] != I8_INIT) {
      if (rank == 0)
        printf("integer(8) was corrupted %ld\n", i8[i]);
      goto end;
    }

    if (r4[i] != R4_INIT) {
      if (rank == 0)
        printf("real(4) was corrupted %f\n", r4[i]);
      goto end;
    }
    if (r8[i] != R8_INIT) {
      if (rank == 0)
        printf("real(8) was corrupted %f\n", r8[i]);
      goto end;
    }
    if (r16[i] != R16_INIT) {
      if (rank == 0)
        printf("real(16) was corrupted %Lf\n", r16[i]);
      goto end;
    }
    if (cp4[i].i != i || cp4[i].r != i) {
      if (rank == 0)
        printf("complex(4) corrupted r=%f, i=%f\n", cp4[i].r, cp4[i].i);
      goto end;
    }
    if (cp8[i].i != i || cp8[i].r != i) {
      if (rank == 0)
        printf("complex(8) corrupted r=%f, i=%f\n", cp8[i].r, cp8[i].i);
      goto end;
    }
    if (cp16[i].i != i || cp16[i].r != i) {
      if (rank == 0)
        printf("complex(16) corrupted r=%Lf, i=%Lf\n", cp16[i].r, cp16[i].i);
      goto end;
    }
  }
  ret = 0;
  end:
  FTI_Finalize();
  MPI_Finalize();
  return ret;
}

File: testing/suites/features/fortran/datatypes/ckpt_primitives.f90
!> Copyright (c) 2017 Leonardo A. Bautista-Gomez
!> All rights reserved
!>
!> @author Alexandre de Limas Santana (alexandre.delimassantana@bsc.es)
!> @date   August, 2020
!> @brief  Checkpoint primitive types and validate FTI meta-data

program checkpoint_primitives
   use FTI
#ifndef MPI_USE_HEADER
   use MPI
#else
   include 'mpif.h'
#endif

   ! Application parameters
   integer(1), parameter :: SIZE = 5
   character*255         :: configfile
   integer, target       :: err, FTI_comm_world, doInitData, rank

   ! These are the variables for testing
   ! The datatypes should map directly to C primitive types
   ! This should be verified in the meta-data files
   character(1), pointer :: c1(:)

   logical(1), pointer   :: l1(:)
   logical(2), pointer   :: l2(:)
   logical(4), pointer   :: l4(:)
   logical(8), pointer   :: l8(:)

   integer(1), pointer   :: i1(:)
   integer(2), pointer   :: i2(:)
   integer(4), pointer   :: i4(:)
   integer(8), pointer   :: i8(:)

   real(4), pointer      :: r4(:)
   real(8), pointer      :: r8(:)
   real(16), pointer     :: r16(:)

   complex(4), pointer   :: cp4(:)
   complex(8), pointer   :: cp8(:)
   complex(16), pointer  :: cp16(:)

   ! Parse program arguments before starting
   if (command_argument_count() .NE. 2) THEN
      write (*, *) 'Expected two arguments: configfile (str), doInitData (int)'
      stop
   endif

   ! Re-use the 'configfile' variable as a char buffer
   call get_command_argument(2, configfile)
   read (configfile, *) doInitData
   call get_command_argument(1, configfile)

   allocate (c1(SIZE))

   allocate (l1(SIZE))
   allocate (l2(SIZE))
   allocate (l4(SIZE))
   allocate (l8(SIZE))

   allocate (i1(SIZE))
   allocate (i2(SIZE))
   allocate (i4(SIZE))
   allocate (i8(SIZE))

   allocate (r4(SIZE))
   allocate (r8(SIZE))
   allocate (r16(SIZE))

   allocate (cp4(SIZE))
   allocate (cp8(SIZE))
   allocate (cp16(SIZE))

   call MPI_Init(err)
   FTI_comm_world = MPI_COMM_WORLD
   call FTI_Init(configfile, FTI_comm_world, err)
   call MPI_Comm_rank(FTI_comm_world, rank, err)

   ! Conditionally initialize the data
   ! Any initialization pattern will do as long as
   ! the verification logic mimics the patterns
   if (doInitData /= 0) then
      if (rank == 0) then
         print *, "Application initializes its own data"
      end if

      do i = 1, SIZE
         c1(i) = 'A'

         l1(i) = .true.
         l2(i) = .true.
         l4(i) = .true.
         l8(i) = .true.

         i1(i) = 127 - (i-1)
         i2(i) = 32767 - (i-1)
         i4(i) = 2147483647 - (i-1)
         i8(i) = 2147483647 + (i-1)

         r4(i) = 10 + (i-1)
         r8(i) = 20 + (i-1)
         r16(i) = 30 + (i-1)

         cp4(i) = CMPLX(i-1, i-1)
         cp8(i) = CMPLX(i-1, i-1)
         cp16(i) = CMPLX(i-1, i-1)
      end do
   else if (rank == 0) then
      print *, "Application initializes its data with using checkpoints"
   end if

   call FTI_Protect(0, c1, err)

   call FTI_Protect(10, l1, err)
   call FTI_Protect(11, l2, err)
   call FTI_Protect(12, l4, err)
   call FTI_Protect(13, l8, err)

   call FTI_Protect(20, i1, err)
   call FTI_Protect(21, i2, err)
   call FTI_Protect(22, i4, err)
   call FTI_Protect(23, i8, err)

   call FTI_Protect(30, r4, err)
   call FTI_Protect(31, r8, err)
   call FTI_Protect(32, r16, err)

   call FTI_Protect(40, cp4, err)
   call FTI_Protect(41, cp8, err)
   call FTI_Protect(42, cp16, err)

   ! If the application initialized the data:
   ! 1) Checkpoint the data
   ! 2) Simulate a crash
   if (doInitData /= 0) then
      call FTI_Checkpoint(1, 1, err)
      call MPI_Finalize(err)
      call exit(0)
   else
      call FTI_Recover(err)
      if (err /= 0) then
         call exit(1)
      end if
   end if

   ! Test if data is initialized as it should
   err = 0
   do i = 1, SIZE
      if (c1(i) /= 'A') then
         if (rank == 0) then
            print *, "character(1) was corrupted: ", c1(i)
         end if
         call exit(1)
      end if

      if (l1(i) .neqv. .true.) then
         if (rank == 0) then
            print *, "logical(1) was corrupted: ", l1(i)
         end if
         call exit(1)
      end if
      if (l2(i) .neqv. .true.) then
         if (rank == 0) then
            print *, "logical(2) was corrupted: ", l2(i)
         end if
         call exit(1)
      end if
      if (l4(i) .neqv. .true.) then
         if (rank == 0) then
            print *, "logical(4) was corrupted: ", l4(i)
         end if
         call exit(1)
      end if
      if (l8(i) .neqv. .true.) then
         if (rank == 0) then
            print *, "logical(8) was corrupted: ", l8(i)
         end if
         call exit(1)
      end if

      if (i1(i) /= 127 - (i - 1)) then
         if (rank == 0) then
            print *, "integer(1) was corrupted: ", i1(i)
         end if
         call exit(1)
      end if
      if (i2(i) /= 32767 - (i - 1)) then
         if (rank == 0) then
            print *, "integer(2) was corrupted: ", i2(i)
         end if
         call exit(1)
      end if
      if (i4(i) /= 2147483647 - (i - 1)) then
         if (rank == 0) then
            print *, "integer(4) was corrupted: ", i4(i)
         end if
         call exit(1)
      end if
      if (i8(i) /= 2147483647 + (i - 1)) then
         if (rank == 0) then
            print *, "integer(8) was corrupted: ", i8(i)
         end if
         call exit(1)
      end if

      if (r4(i) /= 10 + (i - 1)) then
         if (rank == 0) then
            print *, "real(8) was corrupted: ", r4(i)
         end if
         call exit(1)
      end if
      if (r8(i) /= 20 + (i - 1)) then
         if (rank == 0) then
            print *, "real(8) was corrupted: ", r8(i)
         end if
         call exit(1)
      end if
      if (r16(i) /= 30 + (i - 1)) then
         if (rank == 0) then
            print *, "real(16) was corrupted: ", r16(i)
         end if
         call exit(1)
      end if
      if (cp4(i) /= CMPLX(i-1, i-1)) then
         if (rank == 0) then
            print *, "complex(4) was corrupted: ", cp4(i)
         end if
         call exit(1)
      end if
      if (cp8(i) /= CMPLX(i-1, i-1)) then
         if (rank == 0) then
            print *, "complex(8) was corrupted: ", cp8(i)
         end if
         call exit(1)
      end if
      if (cp16(i) /= CMPLX(i-1, i-1)) then
         if (rank == 0) then
            print *, "complex(16) was corrupted: ", cp16(i)
         end if
         call exit(1)
      end if
   end do

   call FTI_Finalize(err)
   call MPI_Finalize(err)
   call exit(err)
endprogram
